README.md
# koa-session-local

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![coverall status][coverall-image]][coverall-url]
[![David deps][david-image]][david-url]
[![David devDeps][david-dev-image]][david-dev-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

>Local session storage for Koa sessions using async/await v1.0.4

This is an implementation of the koa-session storage in local system memory. This is intended to be used as in-memory
session storage for testing/development purposes, as none of the data is persisted to permanent storage.

[![NPM](https://nodei.co/npm/koa-session-local.svg?downloads=true)](https://nodei.co/npm/koa-session-local/)

## Table of Contents

* [Install](#install)
* [Usage](#usage)
    * [Basic](#basic)
* [Options](#options)
    * [Garbage collection](#garbage-collection)
* [License](#license)
* [Contributors](#contributors)

## Install

[npm][]:

```sh
npm install --save koa-session-local
```

[yarn][]:

```sh
yarn add koa-session-local
```

## Usage
`koa-session-local` works with [koa-session](https://github.com/koajs/session) (a simple session middleware for Koa), 
and requires Node >= 7.6 for async/await support.

### Basic

```js
const koa = require('koa');
const session = require('koa-session');
const store = require('koa-session-local');

const app = koa();
app.keys = ['keys', 'keykeys'];
app.use(session({
    store: new store(),
    ...
}));

app.use(async ctx => {
    const { session } = ctx;
    let n = session.views || 0;
    session.views = ++n;
    ctx.body = `${n} view(s)`;
});

app.listen(3000);
```

## Options
The module will use the default, or user provided values for `age` and `maxAge` parameters, as well as verify whether a
session is a rolling session, force save session, or otherwise.

### Garbage collection
Garbage collection is an important feature of many session management implementations in NodeJS and other platforms. In
this module, we're provided options for you to control the frequency and other properties of the garbage collection.

Garbage collection is performed on session.get, so there may be a slight request delay when the session garbage
collection has been initiated.

When a session is loaded, a mathematical check is performed against the probability value to determine if there should 
be a garbage collection cycle. If the generated number is less than the 'probability' value, then the sessions in the 
store are scanned for deletion. If the expiry date for an offer is passed, along with the 'maxlifetime' value, 
the sessions are removed. 

_Note: Using this function with a sufficiently low `maxlifetime` value, in conjunction with other `koa-session` settings 
may remove active or recently expired sessions that are not normally ready for garbage collection._

**Options**

* `gc` (_bool_): Enable or disable session garbage collection (gc)
* `probability` (_float_): The probability for which session gc will be triggered, higher is more frequently
* `maxlifetime` (_int_): The is the maximum lifetime in milliseconds after the session expiry a session can live, before it is 
removed
* `debug` (_fn()_): A reference to a function to generate log information (optional)

**Defaults:**

```js
const defaults = {
    gc: false,
    probability: 0.05, /* 5% chance per session.get call */
    maxlifetime: 60 * 1000 /* 60s in ms */
}
```

**To use:**

Enable `gc` and use the default values with no debugging:

```js
app.use(session({
    store: new store({
         gc: true
     }),
    ...
}));
```

Enable `gc` and configure other properties:

```js
app.use(session({
    store: new store({
         gc: true,
         probability: 0.05, /* 5% change */
         maxlifetime: 60 * 1000, /* 60 seconds */
         debug: console.log /* debug messages will be sent to console.log */
     }),
    ...
}));
```

## License

[MIT](LICENSE) &copy; Justin Mitchell (2019)

## Contributors

[![contrib-jmitchell38488]][contrib-jmitchell38488-url]

[npm-image]: https://img.shields.io/npm/v/koa-session-local.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-session-local
[travis-image]: https://img.shields.io/travis/jmitchell38488/koa-session-local
[travis-url]: https://travis-ci.com/jmitchell38488/koa-session-local
[coverall-image]: https://coveralls.io/repos/github/jmitchell38488/koa-session-local/badge.svg?branch=master
[coverall-url]: https://coveralls.io/github/jmitchell38488/koa-session-local?branch=master
[david-image]: https://img.shields.io/david/jmitchell38488/koa-session-local.svg?style=flat-square
[david-url]: https://david-dm.org/jmitchell38488/koa-session-local
[david-dev-image]: https://img.shields.io/david/dev/jmitchell38488/koa-session-local.svg?style=flat-square&label=devDeps
[david-dev-url]: https://david-dm.org/jmitchell38488/koa-session-local#info=devDependencies
[node-image]: https://img.shields.io/badge/node.js-%3E=_7.6-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/koa-session-local.svg?style=flat-square
[download-url]: https://npmjs.org/koa-session-local
[npm]: https://www.npmjs.com/
[yarn]: https://yarnpkg.com/

[contrib-jmitchell38488]: https://avatars1.githubusercontent.com/u/12840052?s=60
[contrib-jmitchell38488-url]: https://github.com/jmitchell38488


HISTORY.md
### v1.0.4
    * Fixed a race condition where multiple garbage collections could be triggered

### v1.0.3
    * Added configurable garbage collection for expired sessions

### v1.0.2
    * Changed store type from array to Map

### v1.0.1
    * Changed throw from text to instance of 'Error' inside the 'set' method

### v1.0.0
    * Initial commit with complete test coverage


test/store.spec.js
const LocalSessionStore = require('../src');

const chai = require('chai');
const dirtyChai = require('dirty-chai');
const sinon = require('sinon');
const sinonChai = require('sinon-chai');
const assert = require('assert');

const { expect } = chai;

chai.use(dirtyChai);
chai.use(sinonChai);

if (parseFloat(process.version.slice(1)) < 7.6) {
    throw 'Cannot execute tests, node >= 7.6 required';
}

const stubCreator = (key, json, ttl, date = Date.now()) => sinon.stub(LocalSessionStore, 'getNewSessionEntry').returns({
    created: date,
    key: key,
    json: json,
    ttl: ttl
});

describe('suite: Local storage', () => {
    let store;
    let data;
    let optSpy;
    let opts;
    let ttl = 15 * 1000;
    let key1 = 'abc1234567890';
    let key2 = '0987654321cba';
    let key3 = 'foobarasasdas';

    beforeEach(() => {
        store = new LocalSessionStore();
        data = {
            views: 1
        };
        opts = {};
        optSpy = null;
    });

    describe('test: constructor', () => {
        it('Should return 0 sessions', () => {
            expect(store.size).to.equal(0);
        });

        it('should return a session store with destroy, get and set methods', () => {
            expect(store.destroy).to.be.a('function');
            expect(store.get).to.be.a('function');
            expect(store.set).to.be.a('function');
        });

        it('should not throw an error', () => {
            try {
                store = new LocalSessionStore();
                assert(store !== null);
            } catch (e) {
                // truly hacky bs way around this
                assert.equal(false, true);
            }
        });

        it('shold set default gc values', () => {
            assert.equal(store.gc, false);
            assert.equal(store.probability, 0.05);
            assert.equal(store.maxlifetime, 60000);
            assert.equal(store.debug, null);
            assert.equal(store.gcActive, false);
        });

        it('should call _verifyOpts', () => {
            let stub = sinon.stub(LocalSessionStore.prototype, '_verifyOpts').callsFake();
            store = new LocalSessionStore();
            assert(stub.calledOnce);
            stub.restore();
        });

        it('should throw exception for invalid data', () => {
            let opts = 'foo';
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options must be an object, invalid value provided');
            }
        });

        it('should throw exception for invalid gc data', () => {
            let opts = {gc: 'foo'};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.gc must be a boolean');
            }
        });

        it('should throw exception for invalid probability data', () => {
            let opts = {probability: 'foo'};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.probability must be a number');
            }

            opts = {gc: true, probability: 1};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.probability must be a number equal to or less than 1 and greater than 0 when options.gc is enabled');
            }

            opts = {gc: true, probability: 0};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.probability must be a number equal to or less than 1 and greater than 0 when options.gc is enabled');
            }
        });

        it('should throw exception for invalid maxlifetime data', () => {
            let opts = {maxlifetime: 'foo'};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.maxlifetime must be a number');
            }

            opts = {gc: true, maxlifetime: -1};
            try {
                store = new LocalSessionStore(opts);
            } catch (e) {
                assert(e instanceof TypeError);
                assert.equal(e.message, 'options.maxlifetime must be a number greater than 0 when options.gc is enabled');
            }
        });

        it('should not set a debug function', () => {
            assert.equal(store.debug, null);
        });

        it('should set a debug function', () => {
            opts = {debug: console.log};
            store = new LocalSessionStore(opts);

            assert.equal(store.debug, opts.debug);
            assert.equal(typeof store.debug, 'function');
        })
    });

    describe('test: creating a session', () => {
        let creatorStub;

        beforeEach(() => {
            creatorStub = stubCreator(key1, data, ttl);
            store = new LocalSessionStore();
        });

        afterEach(() => {
            creatorStub.restore();
        });

        it('session should create with changed', async () => {
            opts = {
                get changed() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'changed', ['get']);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.calledOnce);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
        });

        it('session should create with rolling', async () => {
            opts = {
                get rolling() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'rolling', ['get']);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.calledOnce);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
        });

        it('session should updated changed', async () => {
            opts = {
                get changed() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'changed', ['get']);

            store.sessions.set(key1, {created: Date.now(), key: key1});
            assert.equal(store.size, 1);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.notCalled);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
        });

        it('session should updated rolling', async () => {
            opts = {
                get rolling() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'rolling', ['get']);

            store.sessions.set(key1, {created: Date.now(), key: key1});
            assert.equal(store.size, 1);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.notCalled);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
        });

        it('session should throw error on create without rolling and changed', async () => {
            opts = {
                get changed() {
                    return false;
                }
            };

            optSpy = sinon.spy(opts, 'changed', ['get']);

            store.set(key1, data, ttl, opts)
                .then(() => {
                    assert(creatorStub.notCalled);
                    assert.equal(optSpy.get.callCount, 1);
                    assert.equal(store.size, 0);
                })
                .catch((e) => {assert.equal(e.message, 'Cannot resolve session')});
        });

        it('session should renew existing session', async () => {
            opts = {
                get renew() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'renew', ['get']);
            let destroySpy = sinon.spy(store, 'destroy');

            store.sessions.set(key1, {created: Date.now(), key: key1});
            assert.equal(store.size, 1);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.calledOnce);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
            assert(destroySpy.calledOnce);
            assert(destroySpy.calledWith(key1));
        });

        it('session should force save an existing session', async () => {
            opts = {
                get force() {
                    return true;
                }
            };

            optSpy = sinon.spy(opts, 'force', ['get']);
            let destroySpy = sinon.spy(store, 'destroy');

            store.sessions.set(key1, {created: Date.now(), key: key1});
            assert.equal(store.size, 1);

            await store.set(key1, data, ttl, opts);
            assert(creatorStub.calledOnce);
            assert.equal(optSpy.get.callCount, 1);
            assert.equal(store.size, 1);
            assert(destroySpy.calledOnce);
            assert(destroySpy.calledWith(key1));
        });
    });

    describe('test: getting a session', () => {
        let getSpy;
        let findSpy;

        beforeEach(() => {
            store = new LocalSessionStore();
            let now = Date.now();
            data = {
                created: now,
                key: key1,
                ttl: ttl,
                json: { views: 1, _expire: now + ttl }
            };

            findSpy = sinon.spy(store.sessions, 'get');
        });

        afterEach(() => {
            data = {};
            store = null;
            getSpy = null;
            findSpy = null;
            Date.now.restore();
        });

        it('should return session that is of a session length expiry', async () => {
            data.json._session = true;
            let getSpy = sinon.spy(Date, 'now');

            store.sessions.set(data.key, data);
            let session = await store.get(key1);

            assert(findSpy.calledOnce);
            assert.equal(JSON.stringify(session), JSON.stringify(data.json));
            assert(getSpy.notCalled);
        });

        it('should return session that has not expired', async () => {
            let getSpy = sinon.spy(Date, 'now');

            store.sessions.set(data.key, data);
            let session = await store.get(key1);

            assert(findSpy.calledOnce);
            assert.equal(JSON.stringify(session), JSON.stringify(data.json));
            assert(getSpy.calledOnce);
        });

        it('should return null for an expired session', async () => {
            let getSpy = sinon.spy(Date, 'now');
            data.json._expire -= 20 * 1000;

            store.sessions.set(data.key, data);
            let session = await store.get(key1);

            assert(findSpy.calledOnce);
            assert.equal(session, null);
            assert(getSpy.calledOnce);
        });

        it('should return null for no session found', async () => {
            let getSpy = sinon.spy(Date, 'now');
            let session = await store.get(key1);

            assert(findSpy.calledOnce);
            assert(getSpy.notCalled);
            assert.equal(session, null);
        });

        it('should call gc function', async () => {
            let getSpy = sinon.spy(Date, 'now');
            let spy = sinon.spy(store, '_performGc');

            store.sessions.set(data.key, data);
            await store.get(key1);
            assert(getSpy.calledOnce);

            spy.restore();
        })
    });

    describe('test: destroying a session', () => {
        let delSpy;

        beforeEach(() => {
            store = new LocalSessionStore();
            delSpy = sinon.spy(store.sessions, 'delete');
            let now = Date.now();
            data = {
                created: now,
                key: key1,
                ttl: ttl,
                json: { views: 1, _expire: now + ttl }
            };
        });

        afterEach(() => {
            data = {};
            store = null;
            delSpy = null;
        });

        it('should destroy found session', async () => {
            store.sessions.set(data.key, data);
            assert.equal(store.size, 1);

            await store.destroy(key1);

            assert(delSpy.calledOnce);
            assert.equal(store.size, 0);
        });

        it('should destroy found session, leave other sessions', async () => {
            store.sessions.set(data.key, data);
            assert.equal(store.size, 1);

            store.sessions.set(key2, Object.assign({}, data, {key: key2}));
            assert.equal(store.size, 2);

            await store.destroy(key1);

            assert(delSpy.calledOnce);
            assert.equal(store.size, 1);
            assert(store.get(key2) !== null);
        });

        it('should not destroy session if not found', async () => {
            data.key = key3;
            store.sessions.set(data.key, data);
            assert.equal(store.size, 1);
            
            await store.destroy(key1);

            assert(delSpy.calledOnce);
            assert.equal(store.size, 1);
        });
    });

    describe('test: it should return a new session template', () => {
        let dateNowStub;
        let dateNow;

        beforeEach(() => {
            store = new LocalSessionStore();
            dateNow = Date.now();
            dateNowStub = sinon.stub(Date, 'now').returns(dateNow);
            data = {
                created: dateNow,
                key: key1,
                json: { views: 1, _expire: dateNow + ttl },
                ttl: ttl
            };
        });

        afterEach(() => {
            dateNowStub.restore();
            store = null;
        });

        it('should return a new session template', () => {
            let session = LocalSessionStore.getNewSessionEntry(data.key, data.json, data.ttl);
            assert(dateNowStub.calledOnce);
            assert.equal(JSON.stringify(session), JSON.stringify(data));
        });
    });

    describe('test: garbage collection', () => {
        let getSpy;
        let findSpy;
        let mathSpy;
        let dateSpy;

        beforeEach(() => {
            let now = Date.now();
            mathSpy = sinon.spy(Math, 'random');
            dateSpy = sinon.spy(Date, 'now');
            data = {
                created: now,
                key: key1,
                ttl: ttl,
                json: { views: 1, _expire: now + ttl }
            };
        });

        afterEach(() => {
            data = {};
            store = null;
            getSpy = null;
            findSpy = null;

            Math.random.restore();
            Date.now.restore();
            mathSpy = null;
            dateSpy = null;
        });

        it('should not perform gc if gc is disabled', () => {
            let opts = {gc: false};
            store = new LocalSessionStore(opts);

            store._performGc();

            assert(mathSpy.notCalled);
        });

        it('should not perform gc if gc is enabled but no sessions', () => {
            let opts = {gc: true};
            store = new LocalSessionStore(opts);

            store._performGc();

            assert(mathSpy.notCalled);
        });

        it('should not perform gc if another active gc is in progress', () => {
            let opts = {gc: true};
            store = new LocalSessionStore(opts);
            store.gcActive = true;

            store._performGc();

            assert(mathSpy.notCalled);
        });

        it('should not perform gc if probability < Math.random', () => {
            let opts = {gc: true, probability: 0.0000001};
            store = new LocalSessionStore(opts);
            store.sessions.set(key1, data);

            let dbgSpy = sinon.spy(store, '_debug');
            store._performGc();

            assert(dbgSpy.notCalled);
            assert(mathSpy.calledOnce);
            assert(dateSpy.notCalled);
        });

        it('should perform gc if probability met, delete 0 sessions', () => {
            let opts = {gc: true, probability: 1};
            store = new LocalSessionStore(opts);
            store.sessions.set(key1, data);

            let dbgSpy = sinon.spy(store, '_debug');
            let delSpy = sinon.spy(store.sessions, 'delete');

            store._performGc();

            assert(dbgSpy.calledTwice);
            assert(mathSpy.calledOnce);
            assert(dateSpy.calledOnce);
            assert(delSpy.notCalled);
            assert.equal(store.size, 1);
        });

        it('should perform gc if probability met, not delete _session:true sessions', () => {
            let opts = {gc: true, probability: 1};
            store = new LocalSessionStore(opts);

            data.json._session = true;
            store.sessions.set(key1, data);

            let dbgSpy = sinon.spy(store, '_debug');
            let delSpy = sinon.spy(store.sessions, 'delete');

            store._performGc();

            assert(dbgSpy.calledTwice);
            assert(mathSpy.calledOnce);
            assert(dateSpy.calledOnce);
            assert(delSpy.notCalled);
            assert.equal(store.size, 1);
            assert.equal(store.gcActive, false);
        });

        it('should perform gc if probability met, delete 1 session', () => {
            let opts = {gc: true, probability: 1, maxlifetime: 1};
            store = new LocalSessionStore(opts);

            data.created -= 90 * 1000;
            data.json._expire = data.created;

            store.sessions.set(key1, data);

            let dbgSpy = sinon.spy(store, '_debug');
            let delSpy = sinon.spy(store.sessions, 'delete');

            store._performGc();

            assert(dbgSpy.calledTwice);
            assert(mathSpy.calledOnce);
            assert(dateSpy.calledOnce);
            assert(delSpy.calledOnce);
            assert.equal(store.size, 0);
            assert.equal(store.gcActive, false);
        });
    });

    describe('test: _debug', () => {
        let logStub;

        beforeEach(() => {
            logStub = sinon.stub();
        });

        afterEach(() => {
            store = null;
            logStub = null;
        });

        it('should not log if not enabled', () => {
            store = new LocalSessionStore();
            store._debug('foo');
            assert(logStub.notCalled);
        });

        it('should log if not enabled', () => {
            store = new LocalSessionStore({debug: logStub});
            store._debug('foo');
            assert(logStub.calledWith('foo'));
            assert(logStub.calledOnce);
        });
    });
});


src/index.js
/**
 * koa-session-local - index.js
 * Copyright(c) 2019
 * MIT Licensed
 *
 * This class should only be used for dev testing purposes. This should _never_ be used in production.
 *
 * Authors:
 *      Justin Mitchell <jmitchell38488@gmail.com> (https://github.com/jmitchell38488)
 */

/**
 * The default options for this store.
 *
 * Garbage collection can be handled through this store by setting the 'gc' flag to true.
 *
 * When a session is loaded, a mathematical check is performed against the probability value to determine if there should
 * be a garbage collection cycle. If the generated number is less than the 'probability' value, then the sessions in
 * the store are scanned for deletion. If the expiry date for an offer is passed, along with the 'maxlifetime' value,
 * the sessions are removed. This may remove active/current sessions.
 * @type {{maxlifetime: number, probability: number, gc: boolean, debug: boolean}}
 */
const defaults = {
    gc: false,
    probability: 0.05,
    maxlifetime: 60 * 1000,
    debug: false
};

/**
 * The LocalSessionStore class is a simple implementation for the Koa Session manager to store sessions in system
 * memory.
 *
 * Sessions are stored locally inside of the instance of the LocalSessionStore with the following structure:
 * {
 *     created: Date.now(),
 *     key: string,
 *     json: {
 *         ... ,
 *         _expire: number,
 *         _maxAge: number,
 *         ?_session: boolean
 *     },
 *     ttl: number
 * }
 *
 * Options that are passed from the koa-session class:
 * {
 *     ?renew: boolean,
 *     ?force: boolean,
 *     ?changed: boolean,
 *     ?maxAge: number,
 *     rolling: boolean
 * }
 */
class LocalSessionStore {

    /**
     * Construct an instance of 'LocalSessionStore' with the options passed. The options are verir
     * @param options
     * @throws TypeError the constructor bubbles any exceptions thrown during initialisation.
     */
    constructor(options) {
        this._verifyOpts(options);

        let opts = Object.assign({}, defaults, (options || {}));

        this.gc = opts.gc;
        this.probability = opts.probability;
        this.maxlifetime = opts.maxlifetime;
        this.debug = (!!opts.debug && typeof opts.debug === 'function') ? opts.debug : null;
        this.gcActive = false;

        // Initialize our sessions container
        this.sessions = new Map();
    }

    /**
     * Helper method to verify the provided options to the constructor match the format of which is required. This is
     * useful for validating session integrity, as well as ensuring the garbage collection (if enabled) is correctly
     * executed.
     * 
     * Options:
     *  - gc (boolean): Enable or disable session garbage collection (gc)
     *  - probability (float): The probability for which session gc will be triggered, higher is more frequently
     *  - maxlifetime (int): This is the maximum lifetime in seconds that a session can live for before it is 
     *                       garbage collected, once it's expiry time is passed
     * options: {
     *     gc: boolean,
     *     probability: float,
     *     maxlifetime: int,
     *     ?debug: fn()
     * }
     * 
     * @param options
     * @throws TypeError Throws when an invalid type is encountered
     * @private
     */
    _verifyOpts(options) {
        if (options !== void 0 && !(!!options && Object.prototype.toString.call(options) === '[object Object]')) {
            throw new TypeError('options must be an object, invalid value provided');
        }

        let opts = Object.assign({}, defaults, (options || {}));

        if (opts.gc === void 0 || (!!opts.gc && typeof opts.gc !== 'boolean')) {
            throw new TypeError('options.gc must be a boolean');
        }

        if (opts.probability === void 0 || (!!opts.probability && typeof opts.probability !== 'number')) {
            throw new TypeError('options.probability must be a number');
        }

        if (opts.gc && (opts.probability > 1 || opts.probability <= 0)) {
            throw new TypeError('options.probability must be a number equal to or less than 1 and greater than 0 when options.gc is enabled');
        }

        if (opts.probability && typeof opts.maxlifetime !== 'number') {
            throw new TypeError('options.maxlifetime must be a number');
        }

        if (opts.gc && opts.maxlifetime < 0) {
            throw new TypeError('options.maxlifetime must be a number greater than 0 when options.gc is enabled');
        }
    }

    /**
     * Helper method to print debug information to the console. To enable, pass a reference to a function in the store
     * constructor.
     * 
     * options: {
     *     ...
     *     debug: fn()
     * }
     * 
     * @param data
     * @private
     */
    _debug(data) {
        if (this.debug) this.debug.call(this.debug, data);
    }

    /**
     * This method is responsible for performing garbage collection on the sessions currently loaded in this store.
     * The method relies on three settings, which can be provided to the constructor:
     *
     *   `gc` enables or disables the garbage collection operation
     *   `probability` is used to determine if garbage collection should be performed
     *   `maxlifetime` is used as a buffer for expired sessions to prevent the removal of any 'current' sessions
     *
     * Depending on how many sessions are currently in the storage engine, there may be a performance hit when the
     * garbage collector is activated, as it will need to scan each session and determine if the expiry date has
     * elapsed, and then remove it from the store.
     *
     * Garbage collection _shold_ only ever be triggered when a session is fetched, as it will remove expired sessions,
     * which may be the one the user is requesting.
     *
     * @private
     */
    _performGc() {
        if (this.gcActive || !this.gc || this.sessions.size === 0) {
            return;
        }

        const num = Math.random();
        if (num > this.probability) {
            return;
        }

        this._debug('performing garbage collection...');
        this.gcActive = true;
        let count = 0;
        const start = new Date().getTime();
        const now = Date.now();

        this.sessions.forEach((s, k) => {
            // don't expire session length sessions, they have to be manually removed by a 'logout'
            if (s && s.json._session) {
                return;
            }

            if (s && now > (s.json._expire + this.maxlifetime)) {
                count++;
                this.sessions.delete(k);
            }
        });

        this.gcActive = false;
        const time = new Date().getTime() - start;
        this._debug(`Finished garbage collection on ${count} session(s) in ${time}ms`);
    }

    /**
     * Implements koa/session.get
     * @param key The session key
     * @param ttl The time-to-live option
     * @param opts Options
     * @returns {Promise<*>} A promise containing the session or null
     */
    // eslint-disable-next-line
    async get(key, ttl, opts) {
        this._performGc();
        const session = this.sessions.get(key);

        // Session-length session does not expire
        if (session && session.json._session) {
            return session.json;
        }

        // Check that the session hasn't already expired
        return (session && session.json._expire > Date.now()) ? session.json : null;
    }

    /**
     * Implements koa/session.set
     * @param key The session key
     * @param json The session data
     * @param ttl The time-to-live option
     * @param opts Options
     * @returns {Promise<void>} An empty promise that resolves on success, or failure on error
     */
    async set(key, json, ttl, opts) {
        // Changed means we just update, or create a new session
        if (opts.changed || opts.rolling) {
            let session = this.sessions.get(key);
            if (!session) {
                this.sessions.set(key, LocalSessionStore.getNewSessionEntry(key, json, ttl));
                return;
            }

            session.ttl = ttl;
            session.json = json;
            return;
        }

        // Renew means we generate a new session with the same key
        // Force saving a new session no matter what
        if (opts.renew || opts.force) {
            await this.destroy(key);
            this.sessions.set(key, LocalSessionStore.getNewSessionEntry(key, json, ttl));
            return;
        }

        // Wait, what happened?!
        throw new Error('Cannot resolve session');
    }

    /**
     * Implements koa/session.destroy
     * @param key The session key
     * @returns {Promise<void>} An empty promise that resolves on success, or failure on error
     */
    async destroy(key) {
        this.sessions.delete(key);
    }

    /**
     * A helper class method to create a new session object
     * @param key
     * @param json
     * @param ttl
     * @returns {{created: number, json: Object, ttl: number, key: string}}
     */
    static getNewSessionEntry(key, json, ttl) {
        return {
            created: Date.now(),
            key: key,
            json: json,
            ttl: ttl
        }
    }

    /**
     * Get the number of sessions that have been created, this does not determine those sessions that need to be pruned
     * @returns {number}
     */
    get size() {
        return this.sessions.size;
    }

}

module.exports = LocalSessionStore;
